File: src/main/java/alfheim/common/entity/boss/ai/flugel/AIBase.java
package alfheim.common.entity.boss.ai.flugel;

import alfheim.common.entity.boss.EntityFlugel;
import net.minecraft.entity.ai.EntityAIBase;

public abstract class AIBase extends EntityAIBase {
	
	EntityFlugel flugel;
	AITask task;
	
	public AIBase(EntityFlugel flugel, AITask task) {
		this.flugel = flugel;
		this.task = task;
		setMutexBits(5);
	}
	
	@Override
	public boolean shouldExecute() {
		return flugel.getHealth() > 0 && flugel.getAITask().equals(task) && flugel.getAITaskTimer() == 0;
	}

	@Override
	public abstract void startExecuting();
	
	public boolean canContinue() {
		if (flugel.getHealth() <= 0 || !flugel.getAITask().equals(task)) return false;
		flugel.setAITaskTimer(flugel.getAITaskTimer() - 1);
		return flugel.getAITaskTimer() > 0;
	}

	@Override
	public abstract boolean continueExecuting();
	
	@Override
	public void resetTask() {
		flugel.setAITaskTimer(0);
		flugel.setAITask(flugel.nextTask());
	}
}

File: src/main/java/alfheim/common/entity/boss/ai/flugel/AIBase.kt
package alfheim.common.entity.boss.ai.flugel

import alfheim.common.entity.boss.EntityFlugel

abstract class AIBase(val flugel: EntityFlugel, val task: AITask) {
	
	open fun shouldExecute(): Boolean {
		return flugel.health > 0
	}
	
	abstract fun startExecuting()
	
	open fun shouldContinue(): Boolean {
		if (flugel.health <= 0 || flugel.aiTask != task) return false
		return --flugel.aiTaskTimer > 0
	}
	
	abstract fun continueExecuting(): Boolean
	
	open fun isInterruptible() = true
	
	open fun endTask() {
		flugel.aiTaskTimer = 0
	}
}

abstract class AIConstantExecutable(val flugel: EntityFlugel) {
	abstract fun execute()
}